TESTS_COMMON/ExhaustibleBlockDevice.h

```diff
@@ -0,0 +1,150 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef MBED_EXHAUSTIBLE_BLOCK_DEVICE_H
+#define MBED_EXHAUSTIBLE_BLOCK_DEVICE_H
+
+#include "BlockDevice.h"
+#include "mbed.h"
+
+
+/** Heap backed block device which simulates failures
+ *
+ * Similar to heap block device but sectors wear out and are no longer programmable
+ * after a configurable number of cycles.
+ *
+ */
+class ExhaustibleBlockDevice : public BlockDevice
+{
+public:
+
+    /** Lifetime of the block device
+     *
+     * @param size      Size of the Block Device in bytes
+     * @param read      Minimum read size required in bytes
+     * @param program   Minimum program size required in bytes
+     * @param erase     Minimum erase size required in bytes
+     */
+    ExhaustibleBlockDevice(bd_size_t size, bd_size_t read=1, bd_size_t program=64, bd_size_t erase=512);
+    virtual ~ExhaustibleBlockDevice();
+
+    /**
+     * Set the number of programming cycles before flash is worn out
+     *
+     *  @param cycles   Program cycles before the device malfunctions or 0 for no limit
+     *  @note The program cycles can only be set before init has been called.
+     */
+    void set_program_cycles(uint32_t cycles);
```

Comment:
Maybe instead of these, a set_program_cycles/get_program_cycles that takes in an address would be useful. That way printing the block device could be done outside of the class.